### PR TITLE
Hide excessive legend keys for vertical legends

### PR DESCRIPTION
**Pull request by michaelnesen**
## What does this implement/fix?

Add support to hide overflowing legend items for vertical legends

- Introduced a `useOverflowLegend` hook to share overflow logic between components 

## Does this close any currently open issues?

Resolves https://github.com/Shopify/core-issues/issues/66615

## What do the changes look like?

https://github.com/Shopify/polaris-viz/assets/90475364/d3cbad03-230d-4461-a046-6dd005db84a4

 
## Storybook link

https://6062ad4a2d14cd0021539c1b-ipjhkjkoxm.chromatic.com/?path=/story/polaris-viz-charts-donutchart--series-colors-up-to-sixteen

- You should see overflow behaviour in vertical legends (i.e DonutChart) 
- Ensure no regression for horizontal legends overflow

### Before merging

- [ ] Check your changes on a variety of [browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers) and devices.

- [ ] Update the Changelog's Unreleased section with your changes.

- [ ] Update relevant documentation, tests, and Storybook.

- [ ] Make sure you're exporting any new shared Components, Types and Utilities from the top level index file of the package

### CODE REVIEW

File: packages/polaris-viz/src/components/DonutChart/components/LegendValues/LegendValues.tsx [hunk 1]
```diff
@@ -1,29 +1,73 @@
...
+    colors: seriesColors,
+    dimensions,
+  });
+
+  const {displayedData, hiddenData} = useVerticalOverflowLegend({
+    data: legendData,
+    height,
+    legendItemDimensions,
+    enableHideOverflow: true,
+  });
```
Comment (michaelnesen) [inline packages/polaris-viz/src/components/DonutChart/components/LegendValues/LegendValues.tsx:66]: This component is the _custom_ legend rendered to show values  + trends in the donut chart, using the `overFlowLegend` hook here to get the same overflow behaviour seen in the shared `LegendContainer` 

https://github.com/Shopify/polaris-viz/assets/90475364/3fd68169-787e-4182-ad8c-7cbf4e7c140a

**Comment by michaelnesen** [comment]
@envex 

Thanks for pointing out the visual regression, I did make a [change](https://github.com/Shopify/polaris-viz/pull/1627/files#diff-b68a22cfc4cb8388a7a91342f0ca5cb446c5da78bc66e311bc9b7a23eb1c2614R122) to only show values in the legend items with prop `showLegendValues` as the [value](https://github.com/Shopify/polaris-viz/pull/1627/files#diff-47ed2c459408b65c4786038afa6c5ef3f48188aa15dfe6cd430b74198e262cd6R62) is now part of the legend data returned from the shared useLegend hook, and [currently on main](https://github.com/Shopify/polaris-viz/blob/main/packages/polaris-viz/src/components/Legend/components/LegendItem/LegendItem.tsx#L120) the legend items show a value whenever it exists. However, besides the chromatic visual tests, I couldn’t find a case where the legend item value was actually being used for any chart 🤔  So i’m inclined to say we can just update the visual tests. 

Looks like values were added to Legends in: https://github.com/Shopify/polaris-viz/pull/1167 but doesn’t seem like any viz/component is actually using the `value` on prop on `LegendItem`

**Review by susiekims — COMMENTED** [review]
Love the hook! 😻

File: packages/polaris-viz/src/components/LegendContainer/LegendContainer.tsx [hunk 2]
```diff
@@ -78,36 +78,27 @@ export function LegendContainer({
...
-    direction,
-  ]);
+  const overflowLegendProps =
+    direction === 'horizontal'
+      ? {
+          direction: 'horizontal' as const,
+          data: allData,
+          enableHideOverflow,
+          legendItemDimensions,
+          width,
```
[the remark this answers is not in the document]
Comment (susiekims) [reply packages/polaris-viz/src/components/LegendContainer/LegendContainer.tsx:88]: Oh weird, I just checked and for the other charts `width` and `dimensions.width` are the same but for donut chart they're different 🤔. Either way it looks like `dimensions.width` is the "correct" one so we can probably just use that instead

[the remark this answers is not in the document]
Comment (michaelnesen) [reply packages/polaris-viz/src/components/LegendContainer/LegendContainer.tsx:88]: @susiekims My mental modal was that `dimensions` is the size of the entire visualization (chart+legend). I had just assumed the width returned from `useLegend` was the width of the legend, but after looking at that logic it seems to actually be the remaining width left in the container for the visualization 🤔 Looks like either way, for horizontal legends the `width` and `dimension.width` will be the same. I can use `dimensions.width` and remove the `width` prop 👍

**Review by susiekims — APPROVED** [review]
Looks great!! 🎉

**Pull request merged**